src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
+        case SetFileProcessingResult::Success:
         {
-            // Check that we set in ZooKeeper node only maximum processed file path.
-            // This check can be useful, when multiple table engines consume in ordered mode.
-            String max_file = getMaxProcessedFile();
-            if (max_file.compare(file_path) <= 0)
-                zookeeper->set(zookeeper_processed_path, file_path);
+            std::lock_guard lock(file_status->metadata_lock);
+            file_status->state = FileStatus::State::Processing;
+
+            file_status->profile_counters.increment(ProfileEvents::S3QueueSetFileProcessingMicroseconds, timer.get());
+            timer.cancel();
+
+            if (!file_status->processing_start_time)
+                file_status->processing_start_time = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+
+            break;
+        }
+        case SetFileProcessingResult::AlreadyProcessed:
+        {
+            std::lock_guard lock(file_status->metadata_lock);
+            file_status->state = FileStatus::State::Processed;
+            break;
+        }
+        case SetFileProcessingResult::AlreadyFailed:
+        {
+            std::lock_guard lock(file_status->metadata_lock);
+            file_status->state = FileStatus::State::Failed;
+            break;
+        }
+        case SetFileProcessingResult::ProcessingByOtherNode:
+        {
+            /// We cannot save any local state here, see comment above.
             break;
         }
     }
-    removeProcessingFile(file_path);
+
+    if (result == SetFileProcessingResult::Success)
+        return std::pair(processing_node_holder, file_status);
+
+    return {};
 }
 
+std::pair<S3QueueFilesMetadata::SetFileProcessingResult,
+          S3QueueFilesMetadata::ProcessingNodeHolderPtr> S3QueueFilesMetadata::trySetFileAsProcessingForUnorderedMode(const std::string & path)
+{
+    /// In one zookeeper transaction do the following:
+    /// 1. check that corresponding persistent nodes do not exist in processed/ and failed/;
+    /// 2. create an ephemenral node in /processing if it does not exist;
+    /// Return corresponding status if any of the step failed.
+
+    const auto node_name = getNodeName(path);
+    const auto zk_client = getZooKeeper();
+    auto node_metadata = createNodeMetadata(path);
+    node_metadata.processing_id = getRandomASCIIString(10);
 
-bool S3QueueFilesMetadata::setFileFailed(const String & file_path, const String & exception_message)
+    Coordination::Requests requests;
+
+    requests.push_back(zkutil::makeCreateRequest(zookeeper_processed_path / node_name, "", zkutil::CreateMode::Persistent));
+    requests.push_back(zkutil::makeRemoveRequest(zookeeper_processed_path / node_name, -1));
+
+    requests.push_back(zkutil::makeCreateRequest(zookeeper_failed_path / node_name, "", zkutil::CreateMode::Persistent));
+    requests.push_back(zkutil::makeRemoveRequest(zookeeper_failed_path / node_name, -1));
```

Comment:
We can use `CheckNotExists` to check if path does not exist if this feature in keeper is enabled. We have function `addCheckNotExistsRequest` for this:
https://github.com/ClickHouse/ClickHouse/blob/eeff98f3a31ffeb13c6e592efd2059bb053298af/src/Common/ZooKeeper/ZooKeeper.h#L723-L737

Reply:
I used it, but in commit https://github.com/ClickHouse/ClickHouse/pull/54422/commits/0ddee6ad6c2d201534761944ce182ac529194837 I reverted back to not using it. Because I need to return different value depending on `responses`, but with `addCheckNotExistsRequest` responses size differs (depending on feature support), and this added complication on checking responses (as I do two `addCheckNotExistsRequest` and in row and then some other request) . So I decided not to use this method as it makes it more complicated to further process the result.